core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -100,26 +102,71 @@ class CloudSqlInstance {
     }
 
     synchronized (instanceDataGuard) {
-      this.currentInstanceData = executor.submit(this::performRefresh);
+      this.currentInstanceData = this.startRefreshAttempt();
       this.nextInstanceData = currentInstanceData;
     }
   }
 
   /**
-   * Returns the current data related to the instance from {@link #performRefresh()}. May block if
-   * no valid data is currently available.
+   * Returns the current data related to the instance from {@link #startRefreshAttempt()}. May block
+   * if no valid data is currently available. This method is called by an application thread when it
+   * is trying to create a new connection to the database. (It is not called by a
+   * ListeningScheduledExecutorService task.) So it is OK to block waiting for a future to complete.
+   *
+   * <p>When no refresh attempt is in progress, this returns immediately. Otherwise, it waits up to
+   * 30 seconds. If a refresh attempt succeeds, returns immediately at the end of that successful
+   * attempt. If no attempts succeed within the 30 second timeout, throws a RuntimeException with
+   * the exception from the last failed refresh attempt as the cause.
+   *
+   * <p>The behavior of this method has changed from earlier releases. In version 1.14.1 and
+   * earlier, When no refresh attempt is in progress, returns immediately. Otherwise, blocks
+   * application thread until the current refresh attempt finishes. If the refresh attempt succeeds,
+   * this returns the InstanceData. If not, this throws a RuntimeException, while a new refresh
+   * attempt is submitted to the executor in the background.
    */
   private InstanceData getInstanceData(long timeoutMs) {
     ListenableFuture<InstanceData> instanceDataFuture;
     synchronized (instanceDataGuard) {
       instanceDataFuture = currentInstanceData;
+      // If the currentInstanceData has expired, then force refresh (which will balk if a refresh
+      // is already running) and make this and future requests to getInstanceData wait on the
+      // refresh operation to complete.
+      if (instanceDataFuture.isDone()) {
```

Comment:
I think we don't need this block of code (down to line 147).

Here's my reasoning.

If an instance data has an expired cert, the next connection attempt will fail and trigger a force refresh for us. So in effect, we have this functionality elsewhere in the Java Connector.

Reply:
I agree that a failed connection attempt forces a refresh. However, I think we still need this here to improve the
stability of our customer applications and tests. 

This PR makes the contract of this method more strict. This method will block until it has a *valid* InstanceData or throw an exception if the timeout expires. 

The old implementation would allow this bad scenario: 
1. Application request a database connection.
2. getInstanceData() returns an expired certificate.
3. CoreSocketFactory.createSocket() triggers a force refresh and throws an exception back to the application.
4. Application has to handle a failed database connection caused by our connector.

The new implementation now works like this: 
1. Application request a database connection
2. getInstanceData() blocks until the certificate is refreshed
4. Application continues uninterrupted. 

On the balance, I think this bit of code is worth the improved stability.

Reply:
If we still think this is important, let's pull it out into a separate PR. It seems like a separate concern from the bigger threading fixes here.